File: implementations/react/src/patterns/behavioral/memento/care-taker.ts
import type { EditorMemento } from "./memento";
import type { EditorOriginator } from "./originator";

export class EditorCaretaker {
  private originator: EditorOriginator;
  private undoStack: EditorMemento[] = [];
  private redoStack: EditorMemento[] = [];

  constructor(originator: EditorOriginator) {
    this.originator = originator;
    this.checkpoint("Initial");
  }

  checkpoint(_label?: string) {
    this.undoStack.push(this.originator.createMemento());
    this.redoStack = [];
  }

  canUndo() { return this.undoStack.length > 1; } 
  canRedo() { return this.redoStack.length > 0; }

  undo() {
    if (!this.canUndo()) return;
    const current = this.undoStack.pop()!; 
    this.redoStack.push(current);
    const previous = this.undoStack[this.undoStack.length - 1];
    this.originator.setMemento(previous);
  }

  redo() {
    if (!this.canRedo()) return;
    const next = this.redoStack.pop()!;
    this.undoStack.push(next);
    this.originator.setMemento(next);
  }

  history() {
    return this.undoStack.map(m => m.getState());
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/memento/caretaker.dart
import 'memento.dart';
import 'originator.dart';

class EditorCaretaker {
  final EditorOriginator originator;
  final List<EditorMemento> _undoStack = [];
  final List<EditorMemento> _redoStack = [];

  EditorCaretaker(this.originator) {
    checkpoint('Initial');
  }

  void checkpoint([String? _label]) {
    _undoStack.add(originator.createMemento());
    _redoStack.clear();
  }

  bool canUndo() => _undoStack.length > 1;
  bool canRedo() => _redoStack.isNotEmpty;

  void undo() {
    if (!canUndo()) return;
    final current = _undoStack.removeLast();
    _redoStack.add(current);
    final previous = _undoStack.last;
    originator.setMemento(previous);
  }

  void redo() {
    if (!canRedo()) return;
    final next = _redoStack.removeLast();
    _undoStack.add(next);
    originator.setMemento(next);
  }

  List<Map<String, Object>> history() =>
      _undoStack.map((m) => m.getState()).toList(growable: false);
}
